File: apps/hrms/hrms/hr/page/who_is_in/who_is_in.js
// frappe.pages['who-is-in'].on_page_load = function(wrapper) {
// 	var page = frappe.ui.make_app_page({
// 		parent: wrapper,
// 		title: 'Who Is In ?',
// 		single_column: true
// 	});
// }

frappe.pages['who-is-in'].on_page_load = function(wrapper) {
	var page = frappe.ui.make_app_page({
		parent: wrapper,
		title: 'Who Is In ?',
		single_column: true
	});
	me = frappe.who_is_in;
	frappe.who_is_in.make(page);
}

	
frappe.who_is_in = {
	start : 0,
	who_is_in_data:"",
	notYetInRows:"",
	statTable:'',
	leaveModal:'',
	lateCheckEmpRow:'',
	make: function (page) {
		me.page = page;
		me.body = $("<div></div>").appendTo(me.page.main);
		me.run();
	},
	run: function () {
		var me = frappe.who_is_in;
		frappe.call({
			method: "hrms.hr.page.who_is_in.who_is_in.getWhoIsInData",
			args: {
				start: me.start,
			},
			callback: function (response){
				if (response.message && response.message.length > 0)
				{	
					me.who_is_in_data = response.message;
					me.send_data(me.who_is_in_data)
					response.message.forEach(function (data) {
						if (data) {
							// ...................
						}else{
							frappe.show_alert({ message: __("Data Not Found ! "), indicator: "gray" });
						}
					});
				}
				else
				{
					$('#page-who-is-in').addClass('hidden')
				}
			},
		});

		window.onpopstate = function(event) {
			window.location.reload();
		};
	},

	send_data: function (data) 
	{
		// console.log(data)
		me.statTable = this.statisticsTable(data[0])
		me.notYetInRows = this.NotYetInRow(data[0]['checkin_statistics'])
		me.leave_rows = this.leaveRows(data[0]['leave_applies'])
		me.lateCheckEmpRow = this.lateCheckinEmployee(data[0]['checkin_statistics']['late_checkin_employee'])
		$(frappe.render_template("who_is_in")).appendTo(me.page.main);

		$('#NotYetIn').text(`Not Yet In (${data[0]['checkin_statistics']['not_checked_in']})`);
		$('#lateArrival').text(`Late Arrivals (${data[0]['checkin_statistics']['late_checkins']})`);
		$('#LeaveCount').text(`On Leave (${data[0]['leave_applies']['todays_ttl_leave_app']})`);

	},
	NotYetInRow: function (data){
		let returnRow = "";
		let not_checking_employees = data['not_checking_employees'];


		not_checking_employees.forEach((employee,index) => {
			returnRow += `<div class="d-flex justify-content-between align-items-center border-bottom">
				<div>
					<p class="my-1">${employee['employee_name']}</p> 
					<a class="font-weight-bold" href="/app/employee/${employee['emp_id']}"><small>#${employee['emp_id']}</small></a>
				</div>
				<div>
					${employee['expected_checkin_time']}
				</div>
			</div>
			`
		});

		return returnRow;
	},

	lateCheckinEmployee: function(late_emp_data) {
		let returnRow = "";
		late_emp_data.forEach((employee, index) => {
			// Extract time part (HH:mm:ss) from employee['time']
			let checkinTime = employee['time'].split(" ")[1].split(".")[0]; // Splits and removes the milliseconds
	
			returnRow += `
				<div class="d-flex justify-content-between align-items-center border-bottom">
					<div>
						<p class="my-1">${employee['employee_name']}</p> 
						<a class="font-weight-bold" href="/app/employee/${employee['emp_id']}"><small>#${employee['emp_id']}</small></a>
					</div>
					<div>
						${checkinTime}  <!-- Display the formatted time here -->
					</div>
				</div>
			`;
		});
	
		return returnRow;
	},
	
	leaveRows: function (data){
		let returnRow = "";
		let leave_applications = data['leave_application'];
		// console.log(leave_applications);

	
		leave_applications.forEach((leave, index) => {
			let statusClass = leave['status'] === 'Approved' ? 'text-success' : 'text-info';
			// console.log(leave['emp_id']);
	
			// Safely store the leave data in a data attribute
			let leaveDataString = JSON.stringify(leave).replace(/"/g, '&quot;'); 
	
			returnRow += `
				<div class="d-flex justify-content-between border-bottom">
					<div>
						<p class="my-1">${leave['employee_name']}</p> 
						<a class="font-weight-bold" href="/app/employee/${leave['emp_id']}"><small>#${leave['emp_id']}</small></a>
					</div>
					<div class='text-right'>
						<p class="my-1">${leave['total_leave_days']} Day(s)</p>
						<span class="leave_status">Status : <span class="${statusClass}">${leave['status']}</span></span>
	
						<p>
							<span data-leave-info="${leaveDataString}" data-emp-id="${leave['emp_id']}" 
								data-toggle="modal" data-target="#leaveInfo" class="toggle-icon"
								onclick="me.OpenLeaveModal(this)">
								<i class="fa fa-chevron-down"></i>
							</span>
						</p>
					</div>
				</div>
			`;
		});
	
		return returnRow;
	},
	

	statisticsTable :function(s_data){
		stat_data = s_data['checkin_statistics'];
		leave_data = s_data['leave_applies'];

		let statTable = 	`
			<table class="table table-bordered bg-white text-dark">
				<tr class="table-bordered">
					<th colspan="4">
						<span class="h5">Employees Information for ${stat_data['data_date']} </span>
					</th>
				</tr>
				<tr>
					<td>
						<b class="">${parseFloat(stat_data['not_checked_in_percentage'].toFixed(2))}%</b>
						<p>${stat_data['not_checked_in']} Employees Are Not Yet In</p> 
					</td>
					<td>
						<b class="">${parseFloat(stat_data['late_checkins_percentage'].toFixed(2))}%</b>
						<p>${stat_data['late_checkins']} Employees Are Late In</p> 
					</td>
					<td>
						<b class="">${parseFloat(stat_data['on_time_checkins_percentage'].toFixed(2))}%</b>
						<p>${stat_data['on_time_checkins']} Employees Are On Time</p> 
					</td>
					<td>
						<b class="">${parseFloat(leave_data['leave_application_percent_only_for_today'].toFixed(2))}%</b>
						<p>${leave_data['todays_ttl_leave_app']} Employees Are On Leave</p> 
					</td>
				</tr>
			</table>
			`
		return statTable;
	},
	OpenLeaveModal: function(element) {
		let leaveData = element.getAttribute('data-leave-info');
		let modal_data = '';
	
		let decodedLeaveData = leaveData.replace(/&quot;/g, '"');
	
		try {
			// Parse the leave data string back to a JSON object
			let leaveApplication = JSON.parse(decodedLeaveData);
			
			modal_data = `
			<div class="modal fade" id="leaveInfo" tabindex="-1" aria-labelledby="leaveInfoLabel" aria-hidden="true">
				<div class="modal-dialog">
					<div class="modal-content">
						<div class="modal-header">
							<h5 class="modal-title" id="leaveInfoLabel">Leave Info</h5>
							<button type="button" class="close" data-dismiss="modal" aria-label="Close">
								<span aria-hidden="true">&times;</span>
							</button>
						</div>
						<div class="modal-body">
							<table class="table table-bordered">
								<tr>
									<th>Emp ID</th>
									<td>${leaveApplication['emp_id']}</td>
								</tr>
								<tr>
									<th>Emp Name</th>
									<td>${leaveApplication['employee_name']}</td>
								</tr>
								<tr>
									<th>Leave Start Date</th>
									<td>${leaveApplication['from_date']}</td>
								</tr>
								<tr>
									<th>Leave End Date</th>
									<td>${leaveApplication['to_date']}</td>
								</tr>
								<tr>
									<th>Leave Days</th>
									<td>${leaveApplication['total_leave_days']}</td>
								</tr>
								<tr>
									<th>Leave Approver</th>
									<td>${leaveApplication['leave_approver']}</td>
								</tr>
								<tr>
									<th>Leave Type</th>
									<td>${leaveApplication['leave_type']}</td>
								</tr>
							</table>
						</div>
						<div class="modal-footer text-right">
							<button type="button" class="btn btn-primary" data-dismiss="modal">Close</button>
						</div>
					</div>
				</div>
			</div>`;
	
			// Remove existing modal if it exists
			$('#leaveInfo').remove(); 
	
			// Append the modal to the body
			document.body.insertAdjacentHTML('beforeend', modal_data);
	
			// Show the modal using Bootstrap's modal method
			$('#leaveInfo').modal('show');
	
		} catch (error) {
			console.error('Error parsing leave data:', error);
			alert('Failed to open modal: ' + error.message); // Optional alert for user feedback
		}
	},
	
	

	
	
}

File: apps/hrms/hrms/hr/page/who_is_in/who_is_in.py
import frappe
from frappe.query_builder.functions import Count
from datetime import datetime, timedelta
@frappe.whitelist()
def getWhoIsInData():
	employee_checkin_statistics = get_employee_checkin_statistics()
	leave_applies = leave_applications()  # Call the corrected function
	return [
		{
            "checkin_statistics": employee_checkin_statistics,
            "leave_applies": leave_applies,
        }
    ]
import frappe

def get_employee_checkin_statistics():
    # Step 1: Get all active employees
    active_employees = frappe.get_all(
        "Employee",
        filters=[["status", "=", "Active"]],
        fields=[
            "employee_name",
            "name as emp_id",
            "designation",
            "reports_to",
            "default_shift",
        ]
    )
    
    total_employees = len(active_employees)

    # Step 2: Get today's check-ins
    today = frappe.utils.today()
    start_of_day = f"{today} 00:00:00"
    end_of_day = f"{today} 23:59:59"
    formatted_date = frappe.utils.format_date(today, "dd MMM yyyy")
    
    checkin_records = frappe.get_all(
        "Employee Checkin",
        filters=[
            ["log_type", "=", 'IN'],
            ["time", "between", [start_of_day, end_of_day]]
        ],
        fields=[
            "employee as emp_id",
            "time",  # Check-in time
            "shift_start",  # Retrieve shift start time from check-in records
            "shift_end"     # Retrieve shift end time from check-in records if needed
        ]
    )
    
    # Create a dictionary for fast lookup of check-ins
    checked_in_today = {checkin["emp_id"]: checkin for checkin in checkin_records}
    
    # Initialize counters and list for non-checking employees
    not_checked_in = 0
    late_checkins = 0
    on_time_checkins = 0
    late_checkin_employee = []
    not_checking_employees = []  # List to store names of employees who didn't check in

    # Step 3: Iterate over the active employees
    for employee in active_employees:
        emp_id = employee['emp_id']
        emp_name = employee['employee_name']
        emp_shift = employee['default_shift'] 
        
        if emp_id not in checked_in_today:
            not_checked_in += 1  # Employee did not check in today
            shift_info = frappe.get_all("Shift Type", filters=[["name", "=", emp_shift]], fields=["start_time", "end_time"])
            
            if shift_info:
                expected_checkin_time = shift_info[0]['start_time']
                shift_start = expected_checkin_time  # Set shift_start to expected_checkin_time
            else:
                expected_checkin_time = "Set Emp Shift"
                shift_start = None  # Explicitly set shift_start to None when no shift is found
            
            not_checking_employees.append({"employee_name": emp_name, "emp_id": emp_id, "expected_checkin_time": expected_checkin_time})  # Add to the list
        else:
            checkin_record = checked_in_today[emp_id]
            checkin_time = checkin_record['time']
            shift_start = checkin_record['shift_start']  # Get shift start from check-in record

            # Check if shift_start is not None before comparing
            if shift_start is not None and checkin_time > shift_start:
                late_checkins += 1
                late_checkin_employee.append({"employee_name": emp_name, **checkin_record})
            elif shift_start is not None:
                on_time_checkins += 1

    # Step 4: Calculate percentages safely
    not_checked_in_percentage = (not_checked_in / total_employees * 100) if total_employees else 0
    late_checkins_percentage = (late_checkins / total_employees * 100) if total_employees else 0
    on_time_checkins_percentage = (on_time_checkins / total_employees * 100) if total_employees else 0

    # Step 5: Return the results along with non-checking employees
    return {
        "total_employees": total_employees,
        "not_checked_in": not_checked_in,
        "not_checked_in_percentage": not_checked_in_percentage,
        "late_checkin_employee" : late_checkin_employee,
        "late_checkins": late_checkins,
        "late_checkins_percentage": late_checkins_percentage,
        "on_time_checkins": on_time_checkins,
        "on_time_checkins_percentage": on_time_checkins_percentage,
        "not_checking_employees": not_checking_employees,
        "data_date": formatted_date,
    }


def leave_applications():
    # Get today's date and the date for the next 15 days
    today = datetime.today().date()
    next_15_days = today + timedelta(days=15)

    
    leave_application = frappe.get_all(
        "Leave Application",
        filters=[
            ["status", "not in", ["Rejected", "Cancelled"]],
            ["from_date", "<=", next_15_days],  # Leave starts on or before the next 15 days
            ["to_date", ">=", today]  # Corrected operator here
        ],
        fields=[
            "employee as emp_id",
            "employee_name",
            "status",
            "from_date",
            "to_date",
            "total_leave_days",
            "leave_type",
            "leave_approver"
        ],
        order_by="posting_date"
    )
    total_leave_app = len(leave_application)
    # Step 1: Get the total count of active employees
    total_active_employees = frappe.get_all(
        "Employee",
        filters=[["status", "=", "Active"]],
        fields=["name as emp_id", "employee_name"]  # Corrected field alias
    )
    
    total_employees_count = len(total_active_employees)

    # Step 2: Calculate the percentage of employees who submitted leave applications
    employees_with_leave_applications = len(set(app["emp_id"] for app in leave_application))  # Unique employees who applied

    # Calculate overall leave application percentage
    leave_application_percent = (employees_with_leave_applications / total_employees_count * 100) if total_employees_count > 0 else 0

    # Step 3: Calculate leave applications specifically for today
    today_leave_application = frappe.get_all(
        "Leave Application",
        filters=[
            ["status", "not in", ["Rejected", "Cancelled"]],
            ["from_date", "<=", today],
            ["to_date", ">=", today]
        ],
        fields=["employee as emp_id"]
    )
    todays_ttl_leave_app = len(today_leave_application)
    employees_with_today_leave_applications = len(set(app["emp_id"] for app in today_leave_application))  # Unique employees who applied today

    # Calculate today's leave application percentage
    leave_application_percent_only_for_today = (employees_with_today_leave_applications / total_employees_count * 100) if total_employees_count > 0 else 0

    # Return a well-structured dictionary
    return {
        "leave_application": leave_application,
        "leave_application_percent": leave_application_percent,
        "leave_application_percent_only_for_today": leave_application_percent_only_for_today,
        "total_leave_app":total_leave_app,
        "todays_ttl_leave_app" : todays_ttl_leave_app,
    }
